WiserTaskScheduler/WiserTaskScheduler/Modules/Ftps/Services/FtpsService.cs

```diff
@@ -0,0 +1,268 @@
+using WiserTaskScheduler.Modules.Ftps.Enums;
+using WiserTaskScheduler.Modules.Ftps.Interfaces;
+using WiserTaskScheduler.Modules.Ftps.Models;
+using GeeksCoreLibrary.Core.DependencyInjection.Interfaces;
+using GeeksCoreLibrary.Modules.Ftps.Interfaces;
+using GeeksCoreLibrary.Modules.Ftps.Models;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json.Linq;
+
+namespace WiserTaskScheduler.Modules.Ftps.Services;
+
+public class FtpsService : IFtpsService, IActionsService, IScopedService
+{
+    private readonly IBodyService bodyService;
+    private readonly IFtpHandlerFactory ftpHandlerFactory;
+    private readonly ILogService logService;
+    private readonly ILogger<FtpsService> logger;
+
+    public FtpsService(IBodyService bodyService, IFtpHandlerFactory ftpHandlerFactory, ILogService logService, ILogger<FtpsService> logger)
+    {
+        this.bodyService = bodyService;
+        this.ftpHandlerFactory = ftpHandlerFactory;
+        this.logService = logService;
+        this.logger = logger;
+    }
+    
+    /// <inheritdoc />
+    public Task InitializeAsync(ConfigurationModel configuration)
+    {
+        return Task.CompletedTask;
+    }
+
+    /// <inheritdoc />
+    public async Task<JObject> Execute(ActionModel action, JObject resultSets, string configurationServiceName)
+    {
+        var ftpAction = (FtpModel) action;
+        var ftpHandler = ftpHandlerFactory.GetFtpHandler(ftpAction.Type);
+
+        var ftpSettings = new FtpSettings()
+        {
+            User = ftpAction.User,
+            Password = ftpAction.Password,
+            Host = ftpAction.Host,
+            Port = ftpAction.Port,
+            EncryptionMode = ftpAction.EncryptionMode,
+            UsePassive = ftpAction.UsePassive,
+            SshPrivateKeyPassphrase = ftpAction.SshPrivateKeyPassphrase,
+            SshPrivateKeyPath = ftpAction.SshPrivateKeyPath
+        };
+        
+        await ftpHandler.OpenConnectionAsync(ftpSettings);
+
+        JObject result;
+
+        if (ftpAction.SingleAction)
+        {
+            result = await ExecuteFtpAction(ftpAction, ftpHandler, resultSets, ReplacementHelper.EmptyRows, ftpAction.UseResultSet, configurationServiceName);
+        }
+        else
+        {
+            var rows = ResultSetHelper.GetCorrectObject<JArray>(ftpAction.UseResultSet, ReplacementHelper.EmptyRows, resultSets);
+            var jArray = new JArray();
+            
+            // Execute the action for each result in the stated result set.
+            for (var i = 0; i < rows.Count; i++)
+            {
+                var indexRows = new List<int> { i };
+                jArray.Add(await ExecuteFtpAction(ftpAction, ftpHandler, resultSets, indexRows, $"{ftpAction.UseResultSet}[{i}]", configurationServiceName));
+            }
+            
+            result = new JObject
+            {
+                {"Results", jArray}
+            };
+        }
+
+        await ftpHandler.CloseConnectionAsync();
```

Comment:
Kun je dit in een try/finally of using blok zetten? Nu wordt de verbinding nooit gesloten als er iets fout gaat in de code daarboven.

Reply:
Try-finally blok is toegevoegd nadat de verbinding is geopend zodat als het verbinding maken niet lukt er niet ook nog geprobeerd wordt de verbinding te sluiten.